common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+@ThreadSafe
+@Slf4j
+/**
+ * This is a continuation token based async iterator implementation. This class takes a function that when completed will 
+ * have next batch of results with continuation token. 
+ * This class determines when to call the next iteration of function (if all existing results have been exhausted) and 
+ * ensures there is only one outstanding call. 
+ */
+public class ContinuationTokenAsyncIterator<T, U> implements AsyncIterator<U> {
```

Comment:
Please use more specific names for the type args. `T` and `U` are not very indicative.

Also, doesn't checkstyle complain about not having `@param` args for these type params?